Warn if a low-entropy passphrase is used to create a wallet

nbougalis opened the pull request:
@vinniefalco, @scottschurr (and @JoelKatz for crypto)

scottschurr commented on src/ripple/rpc/handlers/WalletPropose.cpp:
```diff
@@ -31,9 +31,30 @@
+    for (auto& f : freq)
+        f.second /= input.length();
+
+    for (auto const& f : freq)
```
Ultra minor thought.  I might be inclined to move the declaration of `se` down to just in front of the second `for` loop -- just in front of its first usage.

scottschurr commented on src/ripple/rpc/handlers/WalletPropose.cpp:
```diff
@@ -31,9 +31,30 @@
+    for (auto const& f : freq)
+        se += (f.second) * log2(f.second);
+
+    return std::floor (-se * input.length());
```
Here I'll ask for my own education.  This code review led me to read up on Shannon Entropy (yes, my education is sadly lacking).  The stuff I looked at does not suggest that the entropy takes on discrete values.  So I'm curious why you take the `floor` of the calculated entropy.

nbougalis replied:
The function might be a little misnamed. Here's the rationale:

I'm trying to figure out how many bits of entropy this password contains. I do this by taking the number of bits per character in the ideal encoding (`-se`) and then multiplying it by the length of the password. This gives us back the number of bits this password could represent in if it used an ideal encoding.

Since I can only have a integral number of bits, I can't deal with something like 123.875348 bits. Additionally, since I'm using the result as an estimate on the entropy, I prefer to be ultra conservative (that is, to round down). This means that, in the worst case, I simply underestimate the entropy and issue a warning for a strong passphrase as opposed to not warning for a passphrase on the "cusp" of what I've defined as strong and weak.

scottschurr replied:
Thanks for the explanation.

scottschurr commented:
Thanks for the education.  Looks good to me.  And it seems like a good idea. :+1:.

seelabs commented on src/ripple/rpc/handlers/WalletPropose.cpp:
```diff
@@ -31,9 +31,30 @@
 namespace ripple {
 
+double
+shannon_entropy (std::string const& input)
```
If we look at cases where each letter has an equal probability, this calculates:

S*C*lg(S/C), where S=string length and C is the count.

(1) If use unique letters, C=1 and we get S*lg(S).
(2) If we use two letters, C=S/2, and we get S^2*lg(S)/2.

Intuitively, I expect case (1) to have larger entropy than case (2). But I need about 20 characters to get 80 bits from case (1) and only about 13 to get 80 bits from case (2).

Is that right?

Allows 13 alternating "ababa.." to pass this test seems like too low a bar.

(Hopefully I didn't make a silly math mistake that invalidates above. Even so, we should still calculate the shortest allowable pass work with this definition of entropy.)

scottschurr replied:
Good for you, @seelabs.  With your prod I got off my butt and tried out the function here: http://melpon.org/wandbox/permlink/ccoYfIJP31fXpsYk.  It looks like `shannon_entropy()` does pretty much what we hope for.  A string of 20 unique characters gives an entropy of 86.  A string of identical characters gives a entropy of "-0" :smile:.

vinniefalco replied:
Yeah that "negative zero" will get you every time

seelabs replied:
Looks like y'all saw my comment before I deleted it. Yeah, I misread the code, sorry for the noise.

scottschurr commented:
No sweat.  It's good to ask.  And it was a good reminder to me to try stuff like that out.

nbougalis commented:
I have rebased against the latest develop. I renamed the function `shannon_entropy` to `estimate_entropy` to better document the intent of the code and added some comments. I'm going to add the "pass" label, but feel free to remove it if you spot any issues.

scottschurr commented:
Changes look good to me.  Thanks for adding the comments.  :+1:

JoelKatz commented:
:+1:

nbougalis commented:
Merged as ab8102f927e7db5fee19b453206249f446ab9c70
